chainer/training/trainer.py

```diff
@@ -0,0 +1,285 @@
+import collections
+import os
+
+import six
+
+from chainer import reporter as reporter_module
+from chainer.training import extension as extension_module
+from chainer.training import trigger as trigger_module
+
+
+class _ExtensionEntry(object):
+
+    def __init__(self, extension, priority, trigger, invoke_before_training):
+        self.extension = extension
+        self.trigger = trigger
+        self.invoke_before_training = invoke_before_training
+        self.priority = priority
+
+
+class Trainer(object):
+
+    """The standard training loop in Chainer.
+
+    Trainer is an implementation of a training loop. Users can invoke the
+    training by calling the :meth:`run` method.
+
+    Each iteration of the training loop proceeds as follows.
+
+    - Update of the parameters. It includes the mini-batch loading, forward
+      and backward computations, and an execution of the update formula.
+      These are all done by the update object held by the trainer.
+    - Invocation of trainer extensions in the descending order of their
+      priorities. A trigger object is attached to each extension, and it
+      descides at each iteration whether the extension should be executed.
+      Trigger objects are callable objects that take the trainer object as the
+      argument and return a boolean value indicating whether the extension
+      should be called or not.
+
+    Extensions are callable objects that take the trainer object as the
+    argument. There are two ways to define custom extensions: inheriting the
+    :class:`Extension` class, and decorating functions by
+    :func:`make_extension`. See :class:`Extension` for more details on custom
+    extensions.
+
+    Users can register extensions to the trainer by calling the :meth:`extend`
+    method, where some configurations can be added.
+
+    - Trigger of an extension is a callable object that takes the trainer
```

Comment:
Explanation of `trigger` is duplicated. The former one is [here](https://github.com/pfnet/chainer/pull/1285/files#diff-768d99d739390d9a9a936299d6fa7fceR35)